### DISCUSSION TITLE
Implement a trait for `ArrayD` that should work only for 2 dimensional arrays

### DISCUSSION THREAD

**Discussion by daniellga** [rust-ndarray/ndarray#1317] [General]
Using ArrayD here is a must, but I have to implement a trait that should only work for Array2 types, is checking the dimensions in every function the way to go? I am not sure If I could do some trait bounds in order to eliminate checking the dimensions in every function implemented.

Also, is there any performance gain in converting to Array2 before calling other `ArrayBase` functions in this case?

```
use ndarray::ArrayD;

trait T {
    fn a(&self);
    fn b(&self);
}

impl<A> T for ArrayD<A> {
    fn a(&self) {
        assert!(self.ndim() == 2);
        todo!();
    }
    fn b(&self) {
        assert!(self.ndim() == 2);
        todo!();
    }
}
```

**Comment by bluss** [comment]
> is there any performance gain in converting to Array2 before calling other ArrayBase

Normally yes

**Comment by nilgoyette** [comment]
AFAIK, there's not much do do. Either you
- do your `assert!(self.ndim() == 2);` at the start of all your functions
- implement your trait only for `Array2` and cast to `Array2`
